### PR TITLE
[Merged by Bors] - feat(GroupTheory/Archimedean): `LinearOrderedCommGroupWithZero.discrete_or_denselyOrdered`

### PR DESCRIPTION
**Pull request by pechersky**
---

- [x] depends on: #15824 
- [x] depends on: #16615
- [x] depends on: #16616
- [x] depends on: #16873

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/GroupTheory/ArchimedeanDensely.lean [hunk 1]
```diff
@@ -210,3 +210,38 @@ lemma LinearOrderedCommGroup.discrete_or_denselyOrdered :
...
+      simp_all [← Units.val_le_val]
+    · intro a b
+      induction a <;> induction b <;>
+      simp [MulEquiv.withZero]
+  · intro H
+    refine ⟨fun x y h ↦ ?_⟩
+    rcases (zero_le' (a := x)).eq_or_lt with rfl|hx
+    · let y' := Units.mk0 y h.ne'
+      have hy' : y = y' := rfl
+      rw [hy']
```
Comment (YaelDillies) [inline Mathlib/GroupTheory/ArchimedeanDensely.lean:238]:
```suggestion
    · lift y to Units your_monoid using proof_that_y_is_a_unit
```

File: Mathlib/Algebra/GroupWithZero/WithZero.lean [hunk 2]
```diff
@@ -266,6 +266,43 @@ def unitsWithZeroEquiv : (WithZero α)ˣ ≃* α where
...
+def withZeroUnitsEquiv {G : Type*} [GroupWithZero G]
+    [DecidablePred (fun a : G ↦ a = 0)] :
+    WithZero Gˣ ≃* G where
+  toFun := WithZero.recZeroCoe 0 Units.val
+  invFun a := if h : a = 0 then 0 else (Units.mk0 a h : Gˣ)
+  left_inv := (by induction · <;> simp)
+  right_inv _ := by simp only; split <;> simp_all
+  map_mul' x y := by
+    induction x <;> induction y <;>
+    simp [← WithZero.coe_mul, ← Units.val_mul]
```
Comment (YaelDillies) [inline Mathlib/Algebra/GroupWithZero/WithZero.lean:279]: Why is this still in the diff? Did you forget to merge master?

File: Mathlib/Algebra/Order/Archimedean/Basic.lean [hunk 3]
```diff
@@ -427,3 +427,7 @@ instance (priority := 100) FloorRing.archimedean (α) [LinearOrderedField α] [F
     Archimedean α := by
   rw [archimedean_iff_int_le]
   exact fun x => ⟨⌈x⌉, Int.le_ceil x⟩
+
+instance Units.instMulArchimedean (α) [OrderedCommMonoid α] [MulArchimedean α] :
```
Comment (YaelDillies) [inline Mathlib/Algebra/Order/Archimedean/Basic.lean:431]:
```suggestion
@[to_additive]
instance Units.instMulArchimedean (α) [OrderedCommMonoid α] [MulArchimedean α] :
```

File: Mathlib/GroupTheory/ArchimedeanDensely.lean [hunk 4]
```diff
@@ -210,3 +210,36 @@ lemma LinearOrderedCommGroup.discrete_or_denselyOrdered :
   refine (LinearOrderedAddCommGroup.discrete_or_denselyOrdered (Additive G)).imp ?_ id
   rintro ⟨f, hf⟩
   exact ⟨AddEquiv.toMultiplicative' f, hf⟩
+
+/-- Any nontrivial (has other than 0 and 1) linearly ordered mul-archimedean group with zero is
+either isomorphic (and order-isomorphic) to `ℤₘ₀`, or is densely ordered. -/
+lemma LinearOrderedCommGroupWithZero.discrete_or_denselyOrdered (G : Type*)
+    [LinearOrderedCommGroupWithZero G] [Nontrivial Gˣ] [MulArchimedean G] :
+    Nonempty (G ≃*o WithZero (Multiplicative ℤ)) ∨ DenselyOrdered G := by
```
Comment (YaelDillies) [inline Mathlib/GroupTheory/ArchimedeanDensely.lean:218]: Can you not use the notation here?
```suggestion
    Nonempty (G ≃*o ℤₘ₀) ∨ DenselyOrdered G := by
```

Comment (pechersky) [reply Mathlib/GroupTheory/ArchimedeanDensely.lean:218]: `ℤₘ₀` is defined all the way later in a `Valuation` folder file

Comment (YaelDillies) [reply Mathlib/GroupTheory/ArchimedeanDensely.lean:218]: Uh, maybe that should be fixed? It doesn't need any more than `Algebra.Group.Int`

Comment (pechersky) [reply Mathlib/GroupTheory/ArchimedeanDensely.lean]: Will make separate PR

Comment (pechersky) [reply Mathlib/GroupTheory/ArchimedeanDensely.lean]: #16873

**Comment by YaelDillies** [comment]
Can you please merge master?

**Review by YaelDillies — APPROVED** [review]
Thanks!

maintainer merge

**Comment by semorrison** [comment]
bors merge

**Pull request closed without merging**